intellij/src/main/kotlin/motif/intellij/hierarchy/ScopePropertyHierarchyBrowser.kt

```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright (c) 2018-2019 Uber Technologies, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package motif.intellij.hierarchy
+
+import com.intellij.icons.AllIcons
+import com.intellij.ide.hierarchy.HierarchyBrowserBaseEx
+import com.intellij.ide.hierarchy.HierarchyNodeDescriptor
+import com.intellij.ide.hierarchy.HierarchyTreeStructure
+import com.intellij.ide.hierarchy.JavaHierarchyUtil
+import com.intellij.ide.util.treeView.NodeDescriptor
+import com.intellij.openapi.actionSystem.ActionPlaces
+import com.intellij.openapi.actionSystem.DataKey
+import com.intellij.openapi.actionSystem.DefaultActionGroup
+import com.intellij.openapi.project.Project
+import com.intellij.psi.PsiClass
+import com.intellij.psi.PsiElement
+import com.intellij.psi.PsiElementFactory
+import com.intellij.psi.PsiType
+import com.intellij.psi.presentation.java.ClassPresentationUtil
+import com.intellij.ui.treeStructure.Tree
+import motif.ast.IrType
+import motif.ast.intellij.IntelliJClass
+import motif.ast.intellij.IntelliJType
+import motif.core.ResolvedGraph
+import motif.intellij.MotifProjectComponent
+import motif.intellij.ScopeHierarchyUtils.Companion.isMotifScopeClass
+import motif.intellij.hierarchy.ScopeHierarchyBrowser.Companion.LABEL_GO_NEXT_SCOPE
+import motif.intellij.hierarchy.ScopeHierarchyBrowser.Companion.LABEL_GO_PREVIOUS_SCOPE
+import motif.intellij.hierarchy.descriptor.*
+import motif.models.Scope
+import java.awt.GridLayout
+import java.text.MessageFormat
+import java.util.*
+import javax.swing.JLabel
+import javax.swing.JPanel
+import javax.swing.JTree
+import javax.swing.tree.DefaultMutableTreeNode
+
+/*
+ * UI component used to render scope properties.
+ */
+class ScopePropertyHierarchyBrowser(
+        val project: Project,
+        initialGraph: ResolvedGraph,
+        private val rootElement: PsiElement,
+        private val hierarchyType: PropertyHierarchyType)
+    : HierarchyBrowserBaseEx(project, rootElement), MotifProjectComponent.Listener {
+
+    private var graph: ResolvedGraph = initialGraph
+    private lateinit var legendPanel: JPanel
```

Comment:
`lateinit` is dangerous because it can cause NPEs at runtime. A naive solution is to eagerly instantiate the JPanel inline:

```kotlin
private val legendPanel = JPanel(GridLayout())
```

An even better solution is to use Kotlin's [`lazy {}`](https://kotlinlang.org/docs/reference/delegated-properties.html#lazy) Delegate:

```kotlin
private val legendPanel = lazy { JPanel(GridLayout()) }
```

The use of `lazy` here guarantees that the `JPanel` is only instantiated upon the first access to the property. We should use the same pattern for the legendLabel.

Reply:
Great suggestion. Let's do that.